### PR TITLE
itests: validate responses from ethereum json-rpc api

### PR DESCRIPTION
**Pull request by iand**
## Proposed Changes
Validate responses from following ethereum api calls and fix problems detected (see below for summary of fixes)

 - [x] eth_accounts
 - [x] eth_blockNumber
 - [x] eth_call ✅ **fixed in release/1.20.0 branch**
 - [x] eth_chainId
 - [x] eth_estimateGas
 - [x] eth_feeHistory  ✅ **fixes applied**
 - [x] eth_gasPrice
 - [x] eth_getBalance
 - [x] eth_getBlockByHash   ❌ **https://github.com/filecoin-project/lotus/issues/10106**
 - [x] eth_getBlockByNumber   ❌ **https://github.com/filecoin-project/lotus/issues/10106**
 - [x] eth_getBlockTransactionCountByHash
 - [x] eth_getBlockTransactionCountByNumber
 - [x] eth_getCode 
 - [x] eth_getFilterChanges ✅ **fixed in release/1.20.0 branch**
 - [x] eth_getFilterLogs ✅ **fixed in release/1.20.0 branch**
 - [x] eth_getLogs ✅ **fixed in release/1.20.0 branch**
 - [x] eth_getStorageAt ✅ **fixed in release/1.20.0 branch**
 - [x] eth_getTransactionByBlockHashAndIndex ❌ **https://github.com/filecoin-project/lotus/issues/10106** 
 - [x] eth_getTransactionByBlockNumberAndIndex ❌ **https://github.com/filecoin-project/lotus/issues/10106**
 - [x] eth_getTransactionByHash ❌ **https://github.com/filecoin-project/lotus/issues/10106**
 - [x] eth_getTransactionCount
 - [x] eth_getTransactionReceipt  ✅ **fixes applied**
 - [x] eth_maxPriorityFeePerGas
 - [x] eth_newBlockFilter ✅ **fixes applied**
 - [x] eth_newFilter
 - [x] eth_newPendingTransactionFilter  ✅ **fixes applied**
 - [x] eth_sendRawTransaction
 - [x] eth_uninstallFilter

Changes made to fix conformance:

 -  Added json marshal/unmarshal methods to EthFilterID and EthSubscriptionID
 - Changed `EthFeeHistory.OldestBlock` from `uint64` to `EthUint64`
 - Ensured event log topic parsing was consistent by extracting parseEthTopics
 - cbor decode event log topic values
 - Changed `EthBlock.Extradata` type to `EthBytes`

## Checklist

Before you mark the PR ready for review, please make sure that:

- [ ] Commits have a clear commit message.
- [ ] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [ ] CI is green

### CODE REVIEW

**Review by magik6k — COMMENTED** [review]
Can we somehow avoid pulling all those deps in? According to `scripts/gomod-diff.sh` this is like ~54k lines of diff in dependencies

File: itests/eth_conformance_test.go [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+	manet "github.com/multiformats/go-multiaddr/net"
+	"github.com/stretchr/testify/require"
+
+	"github.com/gregdhill/go-openrpc/parse"
+	"github.com/gregdhill/go-openrpc/types"
+	"github.com/gregdhill/go-openrpc/util"
+)
+
+// TODO generate this using reflection
+type ethAPIRaw struct {
```
[the remark this answers is not in the document]
Comment (iand) [reply itests/eth_conformance_test.go:29]: The return types are interface{} to ensure I get raw json and bypass any json unmarshalling code that might be hiding somethiing

**Comment by iand** [comment]
`eth_getBlockByNumber` validation failures:

```
    eth_conformance_test.go:306: - transactions: Must validate at least one schema (anyOf)
    eth_conformance_test.go:306: - transactions.0: Must validate one and only one schema (oneOf)
    eth_conformance_test.go:306: - transactions.0: accessList is required
    eth_conformance_test.go:306: - transactions.0: yParity is required
    eth_conformance_test.go:306: - extraData: Does not match pattern '^0x[0-9a-f]*$'
    eth_conformance_test.go:309: response did not validate
```

**Comment by iand** [comment]
`eth_getBlockByHash` with full transaction validation failures:

```
    eth_conformance_test.go:306: - extraData: Does not match pattern '^0x[0-9a-f]*$'
    eth_conformance_test.go:306: - transactions: Must validate at least one schema (anyOf)
    eth_conformance_test.go:306: - transactions.0: Must validate one and only one schema (oneOf)
    eth_conformance_test.go:306: - transactions.0: accessList is required
    eth_conformance_test.go:306: - transactions.0: yParity is required
    eth_conformance_test.go:309: response did not validate
```

**Comment by iand** [comment]
`eth_getTransactionByHash` similar to the failures from `eth_getBlockByNumber`

```
    eth_conformance_test.go:338: - (root): Must validate one and only one schema (oneOf)
    eth_conformance_test.go:338: - (root): accessList is required
    eth_conformance_test.go:338: - (root): yParity is required
    eth_conformance_test.go:341: response did not validate
```

**Comment by iand** [comment]
`eth_call` gives a cbor decode failure from decoding `invokeResult.MsgRct.Return` in `EthCall`

```
Error:      	Received unexpected error:
            	expected cbor type 'byte string' in input
```

**Comment by iand** [comment]
With regards to the missing fields on Transactions the openrpc spec gives 3 possible response types for a call such as `eth_getTransactionByHash`:

 1. Signed 1559 Transaction: missing `accessList`, `yParity`
 2. Signed 2930 Transaction: missing `accessList`, `gasPrice`, `yParity`
 3. Signed Legacy Transaction: missing `gasPrice`

See https://www.jsonschemavalidator.net/s/euWIlkpO

I'm not sure which transaction type we are aiming to support

**Comment by mur-me** [comment]
Hi Lotus devs!

Users have found regression for `params` field in the input json after update to the `v1.20.0-hyperspace-0131`:
* `params` became mandatory, but shouldn't - eth-go is broken by `eth_chainId`
* `params` couldn't be null - broken eth_maxPriorityFeePerGas

Previously `params` can be in the following conditions:
* params is optional field - request could be without it 
* params: [] - empty
* params: null 
* params: [ array with something meaningful inside - array, string, object, etc]

Please check this issue

**Review by magik6k — APPROVED** [review]
Looks good, will merge after tests pass (the fevm itest is broken on the release branch so not blocking)

**Comment by Stebalien** [comment]
That schema is a lie. See https://github.com/NethermindEth/nethermind/issues/3276

**Pull request merged**